### ISSUE TITLE
Support `is_cuda`

### ISSUE THREAD

**Issue by kshitij12345** [Lightning-AI/lightning-thunder#138]
Something like the following should work
```python
import thunder
import torch

def foo(x):
    if not x.is_cuda:
        x = x.to('cuda')
    return x * x

x = torch.randn(3, device='cpu')
jit_foo = thunder.jit(foo)
o = jit_foo(x)

print(thunder.last_traces(jit_foo)[-1])
```

Above fails with
```python
  File "/home/kkalambarkar/lightning-thunder/thunder/core/proxies.py", line 1234, in __getattr__
    method: None | Callable = resolve_method(attr, self)
  File "/home/kkalambarkar/lightning-thunder/thunder/core/langctxs.py", line 68, in resolve_method
    method: Callable = ctx.get_method(id, *args, **kwargs)
  File "/home/kkalambarkar/lightning-thunder/thunder/torch/langctx.py", line 40, in get_method
    raise AttributeError(f"The {self.name} language context has no method {id}")
AttributeError: The torch language context has no method is_cuda
```

**Comment by shaharelys** [comment]
Hey @kshitij12345! I'm a new contributor here on Thunder. Would it be okay if I look into this and try to come up with a solution?

Thx,
Shahar

**Comment by kshitij12345** [comment]
@shaharelys Ofcourse, that'd be really great. Do ping here if you require any pointers.

cc: @nikitaved

**Comment by k223kim** [comment]
@kshitij12345 Hi! I am also a new contributor on Thunder :) I had a quick conversation with @shaharelys regarding this issue and he was ok with me taking this issue. Just wanted to let you know that I'll be working on it! (Thank you Shahar!)
Best,
Kaeun

**Comment by k223kim** [comment]
Hello team! I just wanted to check if I am on the right track regarding this issue. I have read the comments of [this issue](https://github.com/Lightning-AI/lightning-thunder/issues/64) and tried to do something similar. Just a heads up that I may make trivial mistakes and would love to learn as I go through this. It'll be awesome if anyone can take a look and provide me some guidance! :)

1.  I have checked torch.is_cuda implementation and noticed a couple of things:
- I noticed that this can be as simple as checking the tensor's device something like the following (from [pytorch](https://github.com/pytorch/pytorch/blob/1ea2f1eaa19176b8b5b778bb317203dc7f4fd3dc/torch/storage.py#L81))
```
    @property
    def is_cuda(self):
        return self.device.type == 'cuda'
```
2. Also, by doing `git grep "def device"` (as this will include some kind of 'device' operation), I have found some code that seems relevant in `core/devices.py` and `clang/__init__.py`.

I am thinking of adding some kind of `is_cuda` property in the above files. Would this be an feasible approach? If so, in which file exactly would be an appropriate place to work at? Thank you for your help in advance! (cc. @kshitij12345, @nikitaved )

**Comment by kshitij12345** [comment]
A thing to note is that this is a property and not a method so we should probably implement it like `ndim`, `numel`, `shape` in TensorProxy.

Ref: https://github.com/Lightning-AI/lightning-thunder/blob/393828fa1dd45b3a9ff695f1bdbf005f5a669e5e/thunder/core/proxies.py#L1177-L1188

Example of `is_cuda` usage:
```python
>>> torch.randn(3, 3).is_cuda
False
>>> torch.randn(3, 3).is_cuda()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'bool' object is not callable

```

**Comment by nikitaved** [comment]
@kshitij12345 , why would we pollute proxies with this info? Proxies are meant to be device agnostic.

**Comment by kshitij12345** [comment]
Yeah, ideally we should update only `torch/__init__.py` with a new `torchsymbol`. I think updating `torchsymbol` to work with properties might be useful in future as well to implement other missing properties if any.

**Comment by k223kim** [comment]
Hi @kshitij12345 ! I am reaching out as I am struggling to make some progress regarding this issue. Specifically, I am trying to include an argument `is_property`(and/or `property_name`) in `torchsymbol` and somehow use that as a trigger to create a property. I believe, I have to modify `class Symbol` 's `__call__` method along with some modifications in `torch/langctx.py`. 
Are there any references that you can point at that might help me to expand `torchsymbol` so that it can work with properties? Or do you have any particular suggestions regarding this implementation?
It seems more intuitive to add this to `TorchProxy`, however, as per the discussion above, I do think this should be included in `torchsymbol`. Would appreciate any help you can provide!

**Comment by kshitij12345** [comment]
@k223kim You are right that we will have to update `torch/langctx.py` and but I don't think we need to update `core/symbol.py`. 

After a cursory look, I think we need to update `torchsymbol` in `torch/__init__.py`
https://github.com/Lightning-AI/lightning-thunder/blob/393828fa1dd45b3a9ff695f1bdbf005f5a669e5e/thunder/torch/__init__.py#L73-L75

And probably add something like `register_property` (similar to `register_method`) to `torch/langctx.py`. I think this will also involve updating `LanguageContext` class to have a notion of property.

https://github.com/Lightning-AI/lightning-thunder/blob/393828fa1dd45b3a9ff695f1bdbf005f5a669e5e/thunder/core/langctxs.py#L17-L19

I think we should keep it simple for now and update TensorProxy. (We can revisit adding this mechanism in future)
cc: @nikitaved for his opinion.

Apologies for the confusion.

**Comment by k223kim** [comment]
Wow, thank you for the warm welcome with such a detailed explanation @mruberry and @kshitij12345! It really helped me understand how `torchsymbol` worked with attributes. I have tried out the both methods that @mruberry mentioned above and I think they are working as expected (so cool!). I will open a PR with the second approach as I found more value in extending thunder's capabilities to support different languages.
